src/main/java/com/rapid7/client/dcerpc/io/PacketOutput.java

```diff
@@ -100,29 +110,48 @@ public void writeString(final String string, final boolean nullTerminate)
 
         if (string != null) {
             writeReferentID();
-            writeInt(maximumChars);
-            writeInt(0);
-            writeInt(currentChars);
-            writeChars(string);
-
-            if (nullTerminate) {
-                writeShort((short) 0);
-            }
-
-            align();
+            writeString(string, maximumChars, currentChars, nullTerminate);
         } else {
             writeNull();
         }
     }
 
     public void writeStringRef(final String string, final boolean nullTerminate)
         throws IOException {
-        if (string != null) {
-            writeReferentID();
-            writeString(string, nullTerminate);
+        writeReferentID();
+        writeString(string, nullTerminate);
+    }
+
+    public void writeString(final String string, final boolean nullTerminate)
+        throws IOException {
+        final int maximumChars;
+        final int currentChars;
+
+        if (string == null) {
+            maximumChars = 1;
+            currentChars = 1;
         } else {
-            writeNull();
+            maximumChars = string.length() + (nullTerminate ? 1 : 0);
+            currentChars = string.length() + (nullTerminate ? 1 : 0);
         }
+
+        writeString(string, maximumChars, currentChars, nullTerminate);
+    }
+
+    private void writeString(final String string, final int maximumChars, final int currentChars, final boolean nullTerminate)
+       throws IOException {
+        writeInt(maximumChars); //max_is (max size)
+        writeInt(0); //min_is (offset)
+        writeInt(currentChars); //size_is (actual size)
+        if (string == null) {
+            writeShort(0);
```

Comment:
This is incorrect.  It's impossible to have a null string at this point (you've already written marshalling code for writing the string).

Reply:
Null is possible when `writeString(final String string, final int maximumChars, final int currentChars, final boolean nullTerminate)` is called by `writeString(final String string, final boolean nullTerminate)` 

When marshalling the `NetprPathCanonicalizeRequest` call; setting the pathName to `null` causes a malformed packet exception; whereas the writeShort codepath ensures we get a valid response.